### DISCUSSION TITLE
How to test CPU+GPU parallel processing using CUDA? Kernel build fails after changing GCCVERSION to 8.%.

### DISCUSSION THREAD

**Discussion by octonout** [OE4T/meta-tegra#833] [Q&A]
```
Yocto release: dunfell
meta-tegra branch: patches-l4t-r32.4
Linux4Tegra release: R32.4.3 and JetPack release: 4.4
Image: B2Qt
```

Hi. I am looking to test CPU+GPU parallel processing on the Jetson Nano. For this I added cuda-toolkit to the IMAGE_INSTALL_append as it seems to include many other cuda related recipes as dependencies to the image. As suggested in the README, I have added the `meta-tegra/contrib` layer and configured the build to use GCC 8 in the local.conf using `GCCVERSION = "8.%"`.

Up till now, the builds have always gone through without a hitch. But now, the do_compile task of `linux-tegra_4.9.bb` fails errors. Please check [this paste](https://pastebin.ubuntu.com/p/3t3CGyz4pk/) for the log.

I have tried `bitbake -c cleanall linux-tegra` and `bitbake -c cleansstate b2qt-embedded-qt6-image` (image recipe). But the error persists.

I am not sure how to proceed further. Also, is adding cuda-toolkit the correct way to test parallel processing? Would it be possible to compile/test Cuda C Programs and Python programs on the target with the cuda-toolkit added? Or do I need to add some other packages as well?

Will really appreciate some help here. Thanks.

**Comment by madisongh** [comment]
The error is:
```
  CC      kernel/bounds.s
cc1: error: -Werror=address-of-packed-member: no option -Waddress-of-packed-member
make[3]: *** [/media/dell/ext4_volume/jetson-nano-build-files/tmp/work-shared/jetson-nano-qspi-sd/kernel-source/./Kbuild:21: kernel/bounds.s] Error 1
make[3]: *** Waiting for unfinished jobs....
```
gcc 8 doesn't support that warning, and there's nothing in the kernel build that adds that flag to compilations - in fact, it disables that warning if the compiler supports it.  So something from one of the other layers in your build setup must be adding that flag to CC or CFLAGS...  you could try dumping the output of `bitbake -e virtual/kernel` to a file and looking for `address-of-packed-member` to see where it might be coming from.

>Also, is adding cuda-toolkit the correct way to test parallel processing?

Probably not, as that recipe just pulls in nvcc and other command-line tools and some basic CUDA runtime support.  I'm not sure what you mean by "test parallel processing," but you could add cuda-samples to your image, which will give you a couple of basic test programs. Beyond that, you'd need either recipes for packages/programs that use CUDA that you could add to your image, or as I think you're trying to do here, create a complete SDK-type image for development on the Nano, so you can download and build other programs directly on the target for testing.

>Would it be possible to compile/test Cuda C Programs and Python programs on the target with the cuda-toolkit added?

You would need to add more than just cuda-toolkit to have a full development environment - you'd also want the complete gcc/g++ toolchain, `cuda-toolkit-dev` and other `-dev` packages so you have the libraries and headers.  And more for Python, I'm just not sure what, off the top of my head.

**Reply by octonout** [reply]
Hey. Thanks for the response. A quick grep of the source code and I was able to find that the option was being added from `${SOURCES}/meta-boot2qt/meta-boot2qt-distro/dynamic-layers/tegra/recipes-kernel/linux/linux-tegra_%.bbappend` in `KCFLAGS` variable. I removed it using my own `linux-tegra_4.9.bbappend`. The build went through fine after cleaning the sstate for gcc and gcc-sanitizers.

> 
> > Also, is adding cuda-toolkit the correct way to test parallel processing?
> 
> Probably not, as that recipe just pulls in nvcc and other command-line tools and some basic CUDA runtime support.

I do not necessarily want to develop on the target. If possible I do not mind developing on the host and cross-compiling via the Yocto build system or Cross Compilation SDK on host. In this case will the cuda-toolkit be sufficient?

> I'm not sure what you mean by "test parallel processing,"

I would actually like to build two programs that essentially do the same resource demanding task. One of these programs will run exclusively on the CPU and the other will be a CUDA program that will run parallelly on the CPU+GPU. I want to do this to gauge the advantage that using GPU for accelerating a program can give.

This CUDA C Program example demonstrates parallel execution: https://www.youtube.com/watch?v=2EbHSCvGFM0 maybe this is what I am looking for. Will have to read up a bit more on CUDA.

Thanks!

**Reply by madisongh** [reply]
> If possible I do not mind developing on the host and cross-compiling via the Yocto build system or Cross Compilation SDK on host. In this case will the cuda-toolkit be sufficient?

If you have recipes to build your CUDA applications, you don't need to add cuda-toolkit to your image at all.  The runtime dependencies on the CUDA libraries will be figured out by the build system and included automatically.  There's a `cuda.bbclass` you can inherit in your recipes to set things up for cross building.  If you use a recent version of CMake for building your application, and use its built-in CUDA language support, that should interoperate with the bbclass.
